platform/server/cpp/src/logger.rs

```diff
@@ -0,0 +1,355 @@
+  fn session_id(&self) -> String {
+    self.strategy.session_id()
+  }
+}
+
+struct MetadataProvider {
+  app_id: String,
+  app_version: String,
+  os: String,
+  os_version: String,
+  locale: String,
+}
+
+impl MetadataProvider {
+  const fn new(
+    app_id: String,
+    app_version: String,
+    os: String,
+    os_version: String,
+    locale: String,
+  ) -> Self {
+    Self {
+      app_id,
+      app_version,
+      os,
+      os_version,
+      locale,
+    }
+  }
+}
+
+impl bd_logger::MetadataProvider for MetadataProvider {
+  fn timestamp(&self) -> anyhow::Result<time::OffsetDateTime> {
+    Ok(time::OffsetDateTime::now_utc())
+  }
+
+  fn fields(&self) -> anyhow::Result<bd_logger::AnnotatedLogFields> {
+    Ok(vec![
+      AnnotatedLogField::new_ootb("app_id".to_string(), self.app_id.clone().into()),
+      AnnotatedLogField::new_ootb("app_version".to_string(), self.app_version.clone().into()),
+      AnnotatedLogField::new_ootb("os".to_string(), self.os.clone().into()),
+      AnnotatedLogField::new_ootb("os_version".to_string(), self.os_version.clone().into()),
+      AnnotatedLogField::new_ootb("_locale".to_string(), self.locale.clone().into()),
+    ])
+  }
+}
+
+struct StaticMetadata {
+  app_id: String,
+  app_version: String,
+  platform: Platform,
+  device_id: String,
+}
+
+impl StaticMetadata {
+  pub const fn new(
+    app_id: String,
+    app_version: String,
+    platform: Platform,
+    device_id: String,
+  ) -> Self {
+    Self {
+      app_id,
+      app_version,
+      platform,
+      device_id,
+    }
+  }
+}
+
+impl bd_metadata::Metadata for StaticMetadata {
+  fn collect_inner(&self) -> HashMap<String, String> {
+    [
+      ("app_version".to_string(), self.app_version.clone()),
+      ("app_id".to_string(), self.app_id.clone()),
+    ]
+    .into()
+  }
+
+  fn sdk_version(&self) -> &'static str {
+    // TODO(fz): Figure out the story for the SDK version.
+    "0.1.0"
+  }
+
+  fn device_id(&self) -> String {
+    self.device_id.clone()
+  }
+
+  fn platform(&self) -> &Platform {
+    &self.platform
+  }
+
+  fn os(&self) -> String {
+    if cfg!(target_os = "macos") {
+      "macos".to_string()
+    } else if cfg!(target_os = "linux") {
+      "linux".to_string()
+    } else if cfg!(target_os = "windows") {
+      "windows".to_string()
+    } else {
+      "unknown".to_string()
+    }
+  }
+}
+
+//
+// Naive in-memory key-storage backed by disk files
+//
+
+pub struct DiskStorage {
+  root: PathBuf,
+  state: parking_lot::Mutex<HashMap<String, String>>,
+}
+
+impl DiskStorage {
+  pub fn new(root: PathBuf) -> Result<Self, std::io::Error> {
+    std::fs::create_dir_all(&root)?;
+
+    Ok(Self {
+      root,
+      state: parking_lot::Mutex::new(HashMap::new()),
+    })
+  }
+
+  fn path(&self, key: &str) -> PathBuf {
+    let mut hasher = DefaultHasher::new();
+    key.hash(&mut hasher);
+    self.root.join(hasher.finish().to_string())
+  }
+}
+
+impl Storage for DiskStorage {
```

Comment:
This is not your doing but this should be an async trait and use async file APIs. I would just add a TODO here. I can look into this when I get back into the SDK code next.